### ISSUE TITLE
optimizer doesn't realize that some types are not possibe after if (value instaceof Float)

### ISSUE THREAD

**Issue by flocsy** [markw65/monkeyc-optimizer#10]
Optimizer doesn't recognize possible type after if:

```
class MyApp extends Application.AppBase {
    function initialize() {
        bug(null);
        bug(1.2);
        bug(1);
    }

    function bug(value as Number or Float or Null) as Number or String or Null {
        if (value instanceof Lang.Float) {
            value = foo(value);
        }
        return value;
    }

    (:inline)
    hidden function foo(value as Float or Null) as String? {
        if (value != null) {
            value = value.format("%.2f");
        }
        return value;
    }
}
```

Without the optimizer it works. With the optimizer it doesn't, because it doesn't realize that the type of value in the last line of bug can't be float, because it could only be Float if we didn't enter the if, but null isn't instance of Lang.Float, so we can only have there either Null or whatever foo returns.

optimizer:
ERROR> fenix6: bug.mc:41:9: Expected $.MyApp.bug to return Null or Number or String but got Null or Number or Float or String

compiler:
ERROR: fenix6: bin/optimized/group045-release/source/source/MyApp.mc:46,4: Object of type 'PolyType<Null or $.Toybox.Lang.Float or $.Toybox.Lang.Number or $.Toybox.Lang.String>' does not match return type 'PolyType<Null or $.Toybox.Lang.Number or $.Toybox.Lang.String>'.

**Comment by markw65** [comment]
I think this is because my optimizer doesn't trust Nulls (or non-Nulls), which is because Garmin declares lots of functions as never returning null, when they do in fact return Null.

As a result, any time the optimizer sees `if (value != null) { ... }` it assumes value might be null, even when the type it can deduce at that point in the program doesn't include null.

Of course, in *this* case, we can prove that value really isn't Null, because we checked that its a Float... but to take advantage of that, I would need to keep track of not just the type, but how I found out that the type didn't include Null.

My long term goal for fixing this is to introduce a PossiblyNull type. All the Toybox functions would then be marked as `<ActualReturnType> or PossiblyNull`. 

Basically you can use `T or PossiblyNull` anywhere T would be allowed. Then if x is `T`, the test `x != null` will be resolved at compile time to false, while if x is `T or PossiblyNull`, the test will remain (but PossiblyNull will be removed on the true path) . But this ends up causing lots of issues - for example we're now claiming that `T or PossiblyNull` is a subtype of `T`...

But if I did that, your example would work, because after inlining the `if (value != null)` would be removed.

I suppose a simple fix for this issue might be to insert a cast to the return type after inlining - but that might introduce other problems. I'll think about it.

**Comment by flocsy** [comment]
It sounds like you're carrying the "maybe null, cause I'm not sure the Garmin docs are correct" to some point, but isn't that a choice your code makes? I mean if I declared the input type of my function is not null (that probably means I don't check for null, and if I would pass the output of a Garmin function that claims not to return null, but in fact does return null then it would crash sometimes) then is there any reason not to believe my code?

**Comment by markw65** [comment]
> but isn't that a choice your code makes?
Yes, of course. But what else can I do?

- If I add `or Null` to the return type of every Toybox function, my type checker will complain about code that compiles cleanly with Garmin's.
- If I just believe the types, I'll remove Null checks that look like they're not needed, but really are.

**Comment by flocsy** [comment]
Actually I'm not sure that the problem is about the Null here. It's about the Float. If I understand correctly then it's complaining about:

function bug(...)  as Number or String or Null {
   if (value instanceof Lang.Float) {
            value = foo(value);
   }
   return value;
}
It thinks that value can be Float as well. But that is not related to wherer foo returns null or not. in fact it does explicitly return String or Null. The problem is rather that it doesn't "deduce" that indeed BEFORE the if value can be Float but after the if it can only be either Null (didn't enter the if) or whatever foo can rerturn, that is String or Null. But it can't be Float after the if.

**Comment by markw65** [comment]
I don't think so. Your example works (if you fix it up):

```
function bug(value as Float or String or Null) as Number or String or Null {
    if (value instanceof Lang.Float) {
        value = 42;
    }
    // Ok. value is Number or String or Null
    return value;
}
```
The problem comes when you add the null test:
```
function bug(value as Float or String or Null) as Number or String or Null {
    if (value instanceof Lang.Float) {
        if (value != null) {
            value = 42;
        }
    }
    // Not ok. value is Float or Number or String or Null
    // (but of course, we know that value != null is always true, so this is
    // actually equivalent to the previous one)
    return value;
}
```

And yes, I hate having to do this. My type checker absolutely knows that `value instanceof Lang.Float` implies `value != null`. It just has to pretend it doesn't to avoid issues caused by Toybox routines that are incorrectly declared as never returning null.

**Comment by flocsy** [comment]
Ok, but then in the inlined code (which is your 2nd example, with the null check inside the Float check) you could still eliminate the null check, as you know that null isn't instanceof Float. This would both decrease code size and eliminate the error message.

**Comment by markw65** [comment]
> you could still eliminate the null check, as you know that null isn't instanceof Float

We're going in circles. As I explained at the beginning, the reason I can't do that is because I would need to keep track of how I know that it's not null. But my type system wasn't designed to do that. If I had realized this was a problem before I started work on it, I could have made it keep track of whether non-null-ness was "guaranteed". But I didn't. And now there are thousands of lines of code making changes to types, and I'd have to find every one and decide whether it preserves the "guaranteed-ness" of the null information.

It's definitely fixable - its just a *lot* of work, and hard to get right.

But it's also not clear that it's the correct fix. The only truly correct fix is to assume that all Toybox functions might return null (possibly with an explicit list of exceptions). That's easy to do, but would force you to add null checks all over the place.

I suppose I could have a default list of exceptions, and allow you to add more in the settings. Honestly, that would be my preferred solution.
